### DISCUSSION TITLE
Mechanism to restart pods running outdated proxies

### DISCUSSION THREAD

**Discussion by hercynium** [istio-ecosystem/sail-operator#363] [Ideas]
Recently I was looking for better ways to manage our Istio upgrades at work, and a real pain point was identifying and restarting all the deployments that were still connected to the old Istio revision after we deployed a new revision and updated the webhook tags.

The sail operator looks really exciting, but [it doesn't (yet) have this feature](https://github.com/istio-ecosystem/sail-operator/issues/74) #74 ... so I decided to try creating it. I'm a beginner at golang, but here's what I came up with: [https://github.com/hercynium/istio-fortsa](https://github.com/hercynium/istio-fortsa)

It's very much a work-in-progress, but feel free to use it for ideas to add this functionality to Sail. Even if it's not useful to the project, I'm always welcome to input on how to make my code better, too - issues, comments, PRs, etc!

**Comment by dgn** [comment]
hey @hercynium! First of all, sorry for the late response. I had a look at the repo, but can you maybe quickly outline how your implementation works? We're looking at simply restarting Deployments when there's a new version, but we need to investigate the details, e.g. how a phased rollout would work

**Reply by hercynium** [reply]
No problem!

There are several moving parts in my implementation.


There are three controllers - one each for watching `mutatingwebhooks`, `namespaces`, and `pods`.

The `mutatingwebhooks` and `namespaces` controllers are very similar. When a change that we're interested in (via the `EventFilter`s) is made to one of those resources, some information is collected and some analysis is done. First, revision-tag info is collected, mapping which istio revision each namespace's pods should be connected to. Next, `proxy-status` data is retrieved from Istio, which has information on which pods are connected to which instances of istiod. Then we determine based on the collected data which pods are not connected to the istiod revision that corresponds to the label on each pod's namespace. Finally, we add a label to each "outdated pod" indicating that it is not connected to the desired revision of istio and thus needs to be restarted. 

The `pods` controller watches for pods with that label. When a pod with the label is reconciled, the following happens: The controller of the pod is determined. This is somewhat messy code IMO, but it seems to work well. The controller can be a `Deployment`, a `DaemonSet`, or a `StatefulSet`. When one of those is found, an annotation is put in the pod template part of the resource's spec. This is the same mechanism used by `kubectl rollout restart` and the effect is exactly the same.

There's additional logic in there to make things run smoothly, but that's how it works, as concisely as I can explain.

One thing to note - the part of the code that gets the `proxy-status` data is really janky. Finding no API to do it, I hijacked the code that makes the CLI command `istioctl proxy-status` work, and modified it so I could access the internal data directly. I'm hoping somebody with more golang experience could do this a whole lot better than I did. :)

**Reply by hercynium** [reply]
SO.... I re-wrote the part that gets proxy-status this weekend. It's still... scary, but it now affords a lot more flexibility in making and processing the XDS request data. ☺

**Reply by luksa** [reply]
@hercynium why use XDS? You can figure which revision a pod connects to by inspecting the `istio.io/rev` annotation, as we do [here](https://github.com/istio-ecosystem/sail-operator/blob/c864dfe00d1597eb78d935ebed85fcd3499811a2/controllers/istiorevision/istiorevision_controller.go#L454) to determine whether a revision is in use.

**Reply by hercynium** [reply]
Before writing fortsa, we had a shell script that called `istioctl proxy-status` then scraped the output and did a bunch of other stuff before terminating pods. So it seemed natural to start with that mechanism, but try to get the data directly in code instead of parsing a CLI command's output.

Thinking about it, perhaps the pod controller could just check all the pods in the cluster for their annotation vs the annotation on the namespace they're in. Right now it only has to look at the ones that were marked as outdated when the mutatingwebhook or namespace controllers detected a change. I'll have to think about it some more on how to trigger a check of pods. I only worry that it could be a big load on the k8s API... but maybe that's not really a big problem? I'm still new to this stuff.

**Reply by luksa** [reply]
@hercynium The operator uses a local cache, so checking the annotations doesn't require any API calls.

**Reply by dgn** [reply]
thanks for the summary @hercynium! I agree with @luksa that using the injected annotation is probably simpler (and it's what we had in mind). do you have any logic to detect whether a migration was successful, and if not, perform rollbacks? another thing I'd like to include are phased rollouts, ie not migrate everything at once but have configurable limits on e.g. how many workloads are restarted at the same timec

**Reply by hercynium** [reply]
There's no logic to detect if the migration was successful, as it's the same logic as running `kubectl rollout restart $deployment_name` from the command-line. The controller could probably have logic to continuously check the status of the migration, similar to how `kubectl rollout status` works, though. As for checking if the new proxy is actually working correctly, the only way I know to do that is... via `istioctl proxy-status`.

For "phased rollouts" I think there would have to be a more complex process than what I've got, which is simply using the rate-limiting built into the operator framework (see: https://github.com/hercynium/istio-fortsa/blob/v0.0.30/internal/util/util.go#L71) and I'm not even sure that's working how I expected as I haven't yet observed it during an upgrade on a large cluster. However, I recently installed it to a medium-sized cluster (about 300 meshed pods) in our test env and I hope to see what happens during our next Istio upgrade in a few weeks.

**Reply by hercynium** [reply]
In case you're interested, I've re-written major portions for Fortsa based on your input... In particular, there's no longer any dependency on Istio XDS code! It seems to work well, and has been running stable in a dev cluster for a few days now where I've switched the Istio tags back and forth between different Istio revs. Each time it does the right thing.

I know I should have actual tests, but that will take more time to figure out how to do properly.
